### ISSUE TITLE
funnel shift related miscompile by ARM64 backend with global isel

### ISSUE THREAD

**Issue by regehr** [llvm/llvm-project#56664]
this function is getting miscompiled:
```llvm
declare i9 @llvm.fshr.i9(i9, i9, i9) #0

define i9 @rotr_zero_shift_guard_swapped(i9 %0, i9 %1) {
  %3 = icmp ne i9 %1, 0
  %4 = call i9 @llvm.fshr.i9(i9 %0, i9 %0, i9 %1)
  %5 = select i1 %3, i9 %4, i9 %0
  ret i9 %5
}

attributes #0 = { nocallback nofree nosync nounwind readnone speculatable willreturn }
```

we can illustrate using this driver:
```c
unsigned rotr_zero_shift_guard_swapped(unsigned, unsigned);

#include <stdio.h>

int main(void) {
  printf("%u\n", 511 & rotr_zero_shift_guard_swapped(50, 273));
}
```

result:
```console
Johns-MacBook-Pro:~ regehr$ llc foo.ll && clang foo.c foo.s && ./a.out
134
Johns-MacBook-Pro:~ regehr$ llc -global-isel foo.ll && clang foo.c foo.s && ./a.out
268
Johns-MacBook-Pro:~ regehr$ 
```

cc @ornata @nunoplopes @ryan-berger @nbushehri @aemerson

**Comment by regehr** [comment]
another trigger for what looks like probably the same bug

```shell
Johns-MacBook-Pro:~ regehr$ ~/llvm-project/for-alive/bin/llc foo.ll && clang foo.s foo.c && ./a.out
2147483648
Johns-MacBook-Pro:~ regehr$ ~/llvm-project/for-alive/bin/llc -global-isel foo.ll && clang foo.s foo.c && ./a.out
8388608
Johns-MacBook-Pro:~ regehr$ cat foo.ll
declare i33 @llvm.fshr.i33(i33, i33, i33) #0

define i33 @f(i33 %0) {
  %2 = call i33 @llvm.fshr.i33(i33 1, i33 1, i33 %0)
  ret i33 %2
}

attributes #0 = { nocallback nofree nosync nounwind readnone speculatable willreturn }
Johns-MacBook-Pro:~ regehr$ cat foo.c
unsigned long f(unsigned long);

#include <stdio.h>

int main(void) {
  printf("%lu\n", 0x1ffffffff & f(4933677476ULL));
}
Johns-MacBook-Pro:~ regehr$ 
```

**Comment by aemerson** [comment]
Should be fixed in 93e3aeb9a84f

**Comment by aemerson** [comment]
@tstellar Tom, I saw there's some automation for cherry-picking onto release branches. Can it handle multiple commits? This fix will depend on a test update in order not to merge conflict: c16fa781f47378a63a515d07c169213573ecd72e
